lib/bats-core/test_functions.bash

```diff
@@ -29,7 +29,37 @@ load() {
   source "${file}"
 }
 
+#########
+#  die  #  Abort with helpful and visible message
+#########
+die() {
+  printf "#/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv\n"  >&2
+  printf "#| FAIL: %s\n" "$1"                                      >&2
+  shift
+  # Any more?
+  for line in "$@"; do
+    printf "#|    > %s\n" "$line"                                  >&2
+  done
+
+  printf "#\\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n" >&2
+  false
+}
+
 run() {
+  local expected_rc=
+
+  # Possible first arguments: '!' = any nonzero, '=n' = exactly n
+  if [[ "$1" = '!' ]]; then
+    expected_rc=-1
+    shift
+  elif [[ "$1" =~ ^=([0-9]|[1-9][0-9]|[12][0-9][0-9])$ ]]; then
```

Comment:
Wouldn't `^=[0-9]+$` suffice? It seems the goal is to restrict the value to 0-255 here, which it already fails at because 299 is matches too. If we really want to check, I would propose to check if `$expected_rc` is out of bounds and print a descriptive error message instead of just silently ignoring the check.

I'd go as far as saying: check the first character, if its `!` ensure that nothing else is there. If its `=` check that only a number follows or give an error. If we need to run a command without checks that starts with `!` or `=` we can add a `--` marker to separate the command from the options.

Reply:
Thanks. I reworked that code to make it more robust, and added tests.